monitor/crawler.py

```diff
@@ -328,17 +332,30 @@ def _measure_top_stakers(self) -> dict:
 
     @collector(label="Staker Confirmation Status")
     def _measure_staker_activity(self) -> dict:
-        confirmed, pending, inactive = self.staking_agent.partition_stakers_by_activity()
-        inactive_without_expired = []
-        for staker in inactive:
-            if self._is_staker_expired(staker_address=staker):
-                continue
-            inactive_without_expired.append(staker)
+        if (not self.__cached_staker_activity
+                or not self.__cached_staker_activity_last_updated
+                or (maya.now() - self.__cached_staker_activity_last_updated).seconds >= self.STAKER_ACTIVITY_CACHE_INVALIDATION_TIMEOUT):
+            # either cache has no value (on initialization) or invalidation timeout so cache needs to be updated
+            confirmed, pending, inactive = self.staking_agent.partition_stakers_by_activity()
+            inactive_without_expired = []
+            for staker in inactive:
+                locked_stake = self.staking_agent.non_withdrawable_stake(staker_address=staker)
+                if locked_stake == 0:
+                    # no active stake, ignore
+                    continue
+                inactive_without_expired.append(staker)
+
+            # update cache
+            self.__cached_staker_activity = confirmed, pending, inactive_without_expired
+            self.__cached_staker_activity_last_updated = maya.now()
+
+        # get data from cache
+        confirmed, pending, inactive = self.__cached_staker_activity
```

Comment:
What do you think about caching only expired and exclude them from `partition_stakers_by_activity`, in that case we will have more actual information but still not so much calls

Reply:
Interesting, but I can't specifically exclude expired from the `partition_stakers_by_activity()` call without modifying the nucypher library - it current returns all three.

Unless you mean when using cached inactive stakers, to instead call `get_all_active_stakers()` for the current period, then separate those into active / pending. Then, when the cache is expired use  `partition_stakers_by_activity()` again...?

Reply:
missed that `partition_stakers_by_activity` is part of main lib, anyway I'm for modifying it if it's not to much